### PR TITLE
Also support single string groups claim

### PR DESCRIPTION
**Pull request by loafoe**
## Description
Allows single string group claims e.g. when one overrides the default `groups` field to point to a string claim

## Motivation and Context
We have a use-case where we point to a specific claim that is a string instead of string array. This allows us to validate by adding multiple `--allowed-group` values

## How Has This Been Tested?
Tested by deploying to our cloud environment and against our OIDC provider and pointing to a string claim

## Checklist:

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: providers/oidc.go [hunk 1]
```diff
@@ -282,6 +282,9 @@ func (p *OIDCProvider) findClaimsFromIDToken(ctx context.Context, idToken *oidc.
 func (p *OIDCProvider) extractGroupsFromRawClaims(rawClaims map[string]interface{}) []string {
 	groups := []string{}
 
+	if group, ok := rawClaims[p.GroupsClaim].(string); ok {
```
Comment (JoelSpeed) [inline providers/oidc.go:285]: I wonder if this should be a switch on the type since we are having multiple options here, something like

```
switch rawClaims[p.GroupsClaim].(type) {
case []string:
  ...
case string:
  ...
}
```

I see we aren't doing this at the moment and are instead casting to a `[]interface{}`, perhaps @stefansedich tried this before and it didn't work? Any ideas?

**Review by JoelSpeed — COMMENTED** [review]
Left a suggestion for possible improvement. Please also make sure there are test cases covering this extraction

File: providers/oidc.go (hunk 1 above)
Comment (loafoe) [reply providers/oidc.go]: @JoelSpeed using case statement now and also added tests

Comment (loafoe) [reply providers/oidc.go]: Looks like I missed some other tests. Looking into it

Comment (NickMeves) [reply providers/oidc.go]: I could've sworn there was some conversation on this `rawGroups, ok := rawClaims[p.GroupsClaim].([]interface{})` -- but I can't seem to find anything on the conversation history anymore.

File: providers/oidc_test.go [hunk 2]
```diff
@@ -373,3 +373,36 @@ func TestOIDCProvider_findVerifiedIdToken(t *testing.T) {
...
+			RawClaims: map[string]interface{}{"groups": []string{"one_group", "another_group"}},
+			Groups:    []string{"one_group", "another_group"},
+		},
+		"StringGroup": {
+			RawClaims: map[string]interface{}{"groups": "one_group"},
+			Groups:    []string{"one_group"},
+		},
+		"NumberGroup": {
+			RawClaims: map[string]interface{}{"groups": 42},
+			Groups:    []string(nil),
```
Comment (NickMeves) [inline providers/oidc_test.go:397]: If a number comes in the JWT claim, I would expect the string form of it to be in the session.

File: providers/oidc_test.go [hunk 3]
```diff
@@ -373,3 +373,36 @@ func TestOIDCProvider_findVerifiedIdToken(t *testing.T) {
 	assert.Equal(t, nil, err)
 	assert.Equal(t, true, verifiedIDToken == nil)
 }
+
+func TestExtractRawgroupsFromClaim(t *testing.T) {
```
Comment (NickMeves) [inline providers/oidc_test.go:377]: NIT: `Test_extractRawGroupsFromClaim` - to match the private method naming.

File: providers/oidc_test.go (hunk 2 above)
Comment (loafoe) [reply providers/oidc_test.go:397]: I'll just remove the testcase, this case is just there to see if the method doesn't panic or blow up

File: providers/oidc.go (hunk 1 above)
Comment (loafoe) [reply providers/oidc.go]: The switch implementation, while explicit, is much longer and duplicates most of the logic in the original

File: providers/oidc_test.go (hunk 2 above)
Comment (NickMeves) [reply providers/oidc_test.go:397]: I think you should have the test case, but you need to assert for `[]string{"42"}` -- I think this is a corner case we might run into (we've been burned by IdPs not following specs many times in the past).

File: providers/oidc.go [hunk 4]
```diff
@@ -282,16 +282,23 @@ func (p *OIDCProvider) findClaimsFromIDToken(ctx context.Context, idToken *oidc.
...
+	switch rawGroups := rawClaims[p.GroupsClaim].(type) {
+	case []string:
+		groups = append(groups, rawGroups...)
+	case *[]string:
+		groups = append(groups, *rawGroups...)
+	case string:
+		groups = append(groups, rawGroups)
+	case *string:
+		groups = append(groups, *rawGroups)
+	case []interface{}:
```
Comment (JoelSpeed) [inline providers/oidc.go:294]: I was expecting us to only need the `[]string` and `string` cases. We should look at how this map is being generated but I had assumed it would be from the json unmarshaller, in which case, we should try to replicate this in the tests and verify which code paths are actually being walked here

Comment (stefansedich) [reply providers/oidc.go:294]: From what I found `rawGroup` was always `[]interface{}` that is why I ended up with the looping and casting, does this new version every hit any of the other cases?

Comment (JoelSpeed) [reply providers/oidc.go:294]: Thanks for chipping in, this is kind of what I was wondering as well. I wonder if there's a better way we can do this by unmarshalling into some clever type that can detect if it's a string or list of strings 🤔

File: providers/oidc_test.go (hunk 2 above)
Comment (loafoe) [reply providers/oidc_test.go]: now supports this corner case

File: providers/oidc.go (hunk 4 above)
Comment (loafoe) [reply providers/oidc.go]: The pointer cases seem highly unlikely yes, I just added them as a reflex, removing them for now

**Comment by Bibob7** [comment]
I think this will not work, because of

```
type OIDCClaims struct {
	rawClaims         map[string]interface{}
	UserID            string
	Subject           string `json:"sub"`
	Verified          *bool  `json:"email_verified"`
	PreferredUsername string `json:"preferred_username"`
	Groups            []string
}
```

If the groups claim has the name `groups` before we even call `extractGroupsFromRawClaims` with `idToken.Claims(&claims);` we already try to unmarshal to `OIDCClaims` (in line 231 of file providers/oidc.go).

In my PR I covered this case: [#839](https://github.com/oauth2-proxy/oauth2-proxy/pull/839)

**Comment by Bibob7** [comment]
@loafoe: Can you add your test case to the testcases in `TestCreateSessionStateFromBearerToken` of providers/oidc_test.go
Then you will see, that the current solution will not work.

To solve this, you have to add an annotation for `Groups`:

```
type OIDCClaims struct {
        [...]
	Groups            []string `json:"-"`
}
```

in providers/oidc.go.

**Comment by NickMeves** [comment]
This related PR for groups parsing merged: https://github.com/oauth2-proxy/oauth2-proxy/pull/839

This needs a rebase to accommodate that logic.

**Comment by NickMeves** [comment]
👋  - Do you have the bandwidth to get this over the finish line? I think this logic is needed before group claim extraction gets released.

**Comment by loafoe** [comment]
Yes, will resolve conflicts tomorrow.

**Comment by loafoe** [comment]
@NickMeves rebased against master

File: providers/oidc.go [hunk 5]
```diff
@@ -285,8 +287,14 @@ func (p *OIDCProvider) findClaimsFromIDToken(ctx context.Context, idToken *oidc.
...
-	rawGroups, ok := rawClaims[p.GroupsClaim].([]interface{})
-	if rawGroups != nil && ok {
+	switch rawGroups := rawClaims[p.GroupsClaim].(type) {
+	case []string:
+		groups = append(groups, rawGroups...)
+	case string:
+		groups = append(groups, rawGroups)
+	case int:
+		groups = append(groups, strconv.Itoa(rawGroups))
+	case []interface{}:
```
Comment (NickMeves) [inline providers/oidc.go:297]: All this should be handled by `formatGroup` since it has the logic to handle string vs non-string (including the int case).

So I think this just needs to handle single vs multi in how it calls `formatGroup`

File: providers/oidc.go [hunk 6]
```diff
@@ -5,14 +5,16 @@ import (
 	"encoding/json"
 	"fmt"
 	"reflect"
+	"strconv"
 	"strings"
 	"time"
 
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/logger"
```
Comment (NickMeves) [inline providers/oidc.go:12]: Let's clean up these imports and get it back together and sorted. I suspect an overeager IDE is to blame 😄

Comment (loafoe) [reply providers/oidc.go]: Yep

File: providers/oidc.go (hunk 5 above)
Comment (loafoe) [reply providers/oidc.go:293]: Looking at `formatGroup` it seems to marshal into a JSON string, so in theory a JSON hashmap string could be returned. Will that work out in the downstream usage of the group values? I removed the `int` check but it looks like all other cases would be handled just as well?

Comment (NickMeves) [reply providers/oidc.go:293]: I think we just want to handle `[]interface{}` vs `interface{}` in the switch and let `formatGroup` do all the work on the underlying data in the list or singleton (string, number, complex object, whatever).

Comment (loafoe) [reply providers/oidc.go:293]: @NickMeves hm I'm kind of lost, since `formatGroup` will return a JSON string should I then go on and parse this?

Comment (NickMeves) [reply providers/oidc.go:293]: `formatGroup` already handles if it is a `string` or not and returns the plain string with no JSON unmarshalling if it is a string: https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/oidc.go#L305

So you don't need to differentiate `string` vs `interface` in your switch, `formatGroup` does that. You just need to handle singular vs list, since the old code assumed everything was a list.

And for non string cases, it will handle number & complex based cases.  You should add back the numeric test you previously had though to make sure the number case does JSON unmarshal correctly to a string through.

**Comment by NickMeves** [comment]
So I actually have a major refactor of OIDC functionality in flight that modifies a lot of these code paths. https://github.com/oauth2-proxy/oauth2-proxy/pull/936

Do you mind if I just fix the logic there directly and an entry in the CHANGELOG for your contribution on this PR mentioning it was fixed in #936?

**Comment by loafoe** [comment]
@NickMeves sure, appreciate it 👍

**Comment by NickMeves** [comment]
> @NickMeves sure, appreciate it 👍

https://github.com/oauth2-proxy/oauth2-proxy/pull/936/commits/ecf1892fa58d8142e063b58d34f11f145faa65ab
